File: go/parse.go
package main

import (
    "fmt"
    "os"
    "log"
    "bufio"
    "regexp"
)

func main() {
    dir, err := os.Open("/Users/gdenslow/hackathon/smallset")
    if ( err != nil ) {
        fmt.Println(err)
        os.Exit(1)
    }
    dir_contents, err := dir.Readdir(-1);
    for _, item := range dir_contents {
        var filepath string = dir.Name() + "/" + item.Name()
        fileInfo, err := os.Stat( filepath )
        if err == nil {
            if ! fileInfo.IsDir() {
                parse_email( filepath );
            }
        }
    }
}

func parse_email( file string ) {
    filehandle, err := os.Open(file)
    if err != nil {
        log.Fatal(err)
    }
    defer filehandle.Close();
    scanner := bufio.NewScanner(filehandle)
        scanner.Split(bufio.ScanLines)
    var this_date string
    var this_from string
    var this_subject string
    date_regexp := regexp.MustCompile("^Date:\\s+(.+)")
    from_regexp := regexp.MustCompile("^From:\\s+(.+)")
    subject_regexp := regexp.MustCompile("^Subject:\\s+(.+)")
    for scanner.Scan() {
        line := scanner.Text()
        if ( this_date == "" ) && date_regexp.MatchString( line ) {
            this_date = ( date_regexp.FindStringSubmatch( line ) )[1]
        }
        if ( this_from == "" ) && from_regexp.MatchString( line ) {
            this_from = ( from_regexp.FindStringSubmatch( line ) )[1]
        }
        if ( this_subject == "" ) && subject_regexp.MatchString( line ) {
            this_subject = ( subject_regexp.FindStringSubmatch( line ) )[1]
        }
    }
    fmt.Println(file + "|" + this_date + "|" + this_from + "|" + this_subject);
}

File: perl/parse.pl
use strict;
use Data::Dumper;

my $dir = '/Users/gdenslow/hackathon/smallset';
my @files;
opendir(D,$dir) || die $!;
while( my $file = readdir D ) {
    if ( -f "$dir/$file" ) {
        push( @files, $file );
    }
}
closedir(D);

#print Dumper \@files;

for my $file ( @files ) {
    my($date_sent,$sender,$subject) = parse_email( $file );
}

sub parse_email {
    my $file = shift;
    open(F,"$dir/$file") || die $!;
    my($date_sent,$sender,$subject);
    while( my $line = <F> ) {
        if ( ! $date_sent && $line =~ /^Date:\s+(.+)/ ) {
            $date_sent = $1;
        }
        if ( ! $sender && $line =~ /^From:\s+(.+)/ ){
            $sender = $1;
        }
        if ( ! $subject && $line =~ /^Subject:\s+(.+)/ ){
            $subject = $1;
        }
    }
    close(F);
    print "$dir/$file|$date_sent|$sender|$subject\n";
}

